From 0b656bf09a53e721c6175e868dedf2904d3f15bc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kai=20L=C3=BCke?= <kailuke@microsoft.com>
Date: Wed, 7 Jul 2021 18:40:52 +0200
Subject: [PATCH 15/20] internal/resource/url: support btrfs as OEM partition
 filesystem

When btrfs is used to fit more content into the partition, mounting
fails because ext4 was hardcoded.
When mounting ext4 fails, try mounting as btrfs.
---
 internal/resource/url.go | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/internal/resource/url.go b/internal/resource/url.go
index f7b855d8..e66707fd 100644
--- a/internal/resource/url.go
+++ b/internal/resource/url.go
@@ -801,8 +801,17 @@ func (f *Fetcher) mountOEM(oemMountPath string) error {
 		},
 		"mounting %q at %q", distro.OEMDevicePath(), oemMountPath,
 	); err != nil {
-		return fmt.Errorf("failed to mount device %q at %q: %v",
+		f.Logger.Err("failed to mount ext4 device %q at %q, trying btrfs: %v",
 			distro.OEMDevicePath(), oemMountPath, err)
+		if err := f.Logger.LogOp(
+			func() error {
+				return syscall.Mount(dev[0], oemMountPath, "btrfs", 0, "")
+			},
+			"mounting %q at %q", distro.OEMDevicePath(), oemMountPath,
+		); err != nil {
+			return fmt.Errorf("failed to mount btrfs device %q at %q: %v",
+				distro.OEMDevicePath(), oemMountPath, err)
+		}
 	}
 
 	return nil
-- 
2.49.1

